Assets/MixedRealityToolkit/Utilities/Rendering/MaterialInstance.cs

```diff
@@ -0,0 +1,306 @@
+            get { return AcquireMaterial(); }
+        }
+
+        /// <summary>
+        /// Returns all the instantiated materials of this object, similar to <see href="https://docs.unity3d.com/ScriptReference/Renderer-materials.html">Renderer.materials</see>.
+        /// </summary>
+        public Material[] Materials
+        {
+            get { return AcquireMaterials(); }
+        }
+
+        private Renderer CachedRenderer
+        {
+            get
+            {
+                if (cachedRenderer == null)
+                {
+                    cachedRenderer = GetComponent<Renderer>();
+                }
+
+                return cachedRenderer;
+            }
+        }
+
+        private Renderer cachedRenderer = null;
+
+        [SerializeField, HideInInspector]
+        private Material[] defaultMaterials = null;
+        private Material[] instanceMaterials = null;
+        private bool materialsInstanced = false;
+        private HashSet<Object> materialOwners = new HashSet<Object>();
+
+        private const string instancePostfix = " (Instance)";
+
+#region MonoBehaviour Implementation
+
+        private void Awake()
+        {
+            if (CachedRenderer != null)
+            {
+                // Cache the default materials if ones do not already exist.
+                if (defaultMaterials == null || defaultMaterials.Length == 0)
+                {
+                    defaultMaterials = CachedRenderer.sharedMaterials;
+                }
+                else if (!materialsInstanced) // Restore the clone to it's initial state.
+                {
+                    CachedRenderer.sharedMaterials = defaultMaterials;
+                }
+            }
+        }
+
+        private void Update()
+        {
+            // If the materials get changed via outside of MaterialInstance.
+            var sharedMaterials = CachedRenderer.sharedMaterials;
+
+            if (!MaterialsMatch(sharedMaterials, instanceMaterials))
+            {
+                // Re-create the material instances.
+                var newDefaultMaterials = new Material[sharedMaterials.Length];
+                var min = Math.Min(newDefaultMaterials.Length, defaultMaterials.Length);
+
+                // Copy the old defaults.
+                for (var i = 0; i < min; ++i)
+                {
+                    newDefaultMaterials[i] = defaultMaterials[i];
+                }
+
+                // Patch in the new defaults.
+                for (var i = 0; i < newDefaultMaterials.Length; ++i)
+                {
+                    var material = sharedMaterials[i];
+
+                    if (!IsInstanceMaterial(material))
+                    {
+                        newDefaultMaterials[i] = material;
+                    }
+                }
+
+                defaultMaterials = newDefaultMaterials;
+                CreateInstances();
+
+                // Notify owners of the change.
+                foreach (var owner in materialOwners)
```

Comment:
I believe guidance is to avoid foreach loops in update() https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/Contributing/CodingGuidelines.html#use-for-instead-of-foreach-when-possible

Reply:
Is that guidance for style or performance? In Unity ~5.3+ the allocation issues have been resolved: https://forum.unity.com/threads/upgraded-c-compiler-on-5-3-5p8.417363/ a nice breakdown too: https://jacksondunstan.com/articles/3805

Reply:
materialOwners are a HashSet, so I'm going to resolve this.